components/layout/fragment.rs

```diff
@@ -1203,6 +924,181 @@ impl Fragment {
+                if self.has_intrinsic_ratio() {
+                    // This approch follows the spirit of cover and contain constraint.
+                    // https://drafts.csswg.org/css-images-3/#cover-contain
+
+                    // First, have two rectangles that keep aspect ratio while may be clamped
+                    // by the contraints;
+                    let (first_isize, first_bsize) = {
+                        let isize = inline_constraint.clamp(intrinsic_inline_size);
+                        let bsize = isize * intrinsic_block_size.0 / intrinsic_inline_size.0;
+                        (isize, bsize)
+                    };
+                    let (second_isize, second_bsize) = {
+                        let bsize = block_constraint.clamp(intrinsic_block_size);
+                        let isize = bsize * intrinsic_inline_size.0 / intrinsic_block_size.0;
+                        (isize, bsize)
+                    };
+                    let (inline_size, block_size) = match (first_isize.cmp(&intrinsic_inline_size) ,
+                                                           second_isize.cmp(&intrinsic_inline_size)) {
+                        (Ordering::Equal, Ordering::Equal) =>
+                            (first_isize, first_bsize),
+                        // When only one rectangle is clamped, use it;
+                        (Ordering::Equal, _) =>
+                            (second_isize, second_bsize),
+                        (_, Ordering::Equal) =>
+                            (first_isize, first_bsize),
+                        // When both rectangles grow (smaller than min sizes),
+                        // Choose the larger one;
+                        (Ordering::Greater, Ordering::Greater) if first_isize > second_isize =>
+                            (first_isize, first_bsize),
+                        (Ordering::Greater, Ordering::Greater) =>
+                            (second_isize, second_bsize),
+                        // When both rectangles shrink (larger than max sizes),
+                        // Choose the smaller one;
+                        (Ordering::Less, Ordering::Less) if first_isize > second_isize =>
+                            (second_isize, second_bsize),
+                        (Ordering::Less, Ordering::Less) =>
+                            (first_isize, first_bsize),
+                        // It does not matter which we choose here, because both sizes
+                        // will be clamped to constraint;
+                        (Ordering::Less, Ordering::Greater) | (Ordering::Greater, Ordering::Less) =>
+                            (first_isize, first_bsize)
+                    };
+                    // Clamp the result and we are done :-)
+                    (inline_constraint.clamp(inline_size), block_constraint.clamp(block_size))
+                } else {
+                    (inline_constraint.clamp(intrinsic_inline_size),
```

Comment:
Hmm... I am not sure about this, use two branch rather than an early return feels a clearer logic.